ballerina/tests/data_binding_test.bal

```diff
@@ -0,0 +1,568 @@
+// Copyright (c) 2022 WSO2 Inc. (//www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// //www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/test;
+
+public type Coord record {
+    int x;
+    int y;
+};
+
+json jsonVal = {x: 1, y: 2};
+xml xmlVal = xml `<book>The Lost World</book>`;
+Coord recordVal = {x: 1, y: 2};
+Coord[] recordArrVal = [{x: 1, y: 2}, {x: 3, y: 4}];
+json[] jsonArr = [{"name":"John","salary":100},{"name":"Jane","salary":200}];
+
+public type Employee record {
+    readonly string name;
+    int salary;
+};
+
+table<Employee> key(name) t = table [
+    { name: "John", salary: 100 },
+    { name: "Jane", salary: 200 }
+];
+
+string errorMessage = "data binding failed";
+
+listener Listener l78 = new(22078);
+
+service /onRecord on l78 {
+    resource function get .() returns Service|UpgradeError {
+        return new WsService78();
+    }
+}
+
+service class WsService78 {
+    *Service;
+    remote isolated function onTextMessage(Caller caller, Coord data) returns Error? {
+        check caller->writeTextMessage(data);
+    }
+}
+
+service /onJson on l78 {
+    resource function get .() returns Service|UpgradeError {
+        return new WsService82();
+    }
+}
+
+service class WsService82 {
+    *Service;
+    remote isolated function onTextMessage(Caller caller, json data) returns Error? {
+        check caller->writeTextMessage(data);
+    }
+}
+
+service /onXml on l78 {
+    resource function get .() returns Service|UpgradeError {
+        return new WsService80();
+    }
+}
+
+service class WsService80 {
+    *Service;
+    remote isolated function onTextMessage(Caller caller, xml data) returns Error? {
+        check caller->writeTextMessage(data);
+    }
+}
+
+service /onRecordArr on l78 {
+    resource function get .() returns Service|UpgradeError {
+        return new WsService81();
+    }
+}
+
+service class WsService81 {
+    *Service;
+    remote isolated function onTextMessage(Caller caller, Coord[] data) returns Error? {
```

Comment:
Do we support `readonly & Coord[]` type?

Reply:
Added readonly support with [71835d4](https://github.com/ballerina-platform/module-ballerina-websocket/pull/541/commits/71835d4e52c3ca2355bf4c6d43db3986186987ba)